ic-agent/src/identity/secp256k1.rs

```diff
@@ -3,107 +3,113 @@ use crate::{export::Principal, Identity, Signature};
 #[cfg(feature = "pem")]
 use crate::identity::error::PemError;
 
-use openssl::{
-    bn::BigNumContext,
-    ec::{EcKey, PointConversionForm},
-    ecdsa::EcdsaSig,
-    error::ErrorStack,
-    nid::Nid,
-    pkey::{Private, Public},
-    sha::sha256,
-};
-use simple_asn1::{
-    oid, to_der, ASN1Block,
-    ASN1Block::{BitString, ObjectIdentifier, Sequence},
+use k256::{
+    ecdsa::{self, signature::Signer, SigningKey, VerifyingKey},
+    elliptic_curve::{sec1::ToEncodedPoint, AlgorithmParameters},
+    pkcs8::{PublicKeyDocument, SubjectPublicKeyInfo},
+    Secp256k1, SecretKey,
 };
+use std::convert::TryInto;
+#[cfg(feature = "pem")]
+use std::{fs::File, io, path::Path};
 
 /// A cryptographic identity based on the Secp256k1 elliptic curve.
 ///
 /// The caller will be represented via [`Principal::self_authenticating`], which contains the SHA-224 hash of the public key.
 #[derive(Clone, Debug)]
 pub struct Secp256k1Identity {
-    private_key: EcKey<Private>,
-    _public_key: EcKey<Public>,
-    der_encoded_public_key: Vec<u8>,
+    private_key: SigningKey,
+    _public_key: VerifyingKey,
+    der_encoded_public_key: PublicKeyDocument,
 }
 
 impl Secp256k1Identity {
     /// Creates an identity from a PEM file. Shorthand for calling `from_pem` with `std::fs::read`.
     #[cfg(feature = "pem")]
-    pub fn from_pem_file<P: AsRef<std::path::Path>>(file_path: P) -> Result<Self, PemError> {
-        Self::from_pem(std::fs::File::open(file_path)?)
+    pub fn from_pem_file<P: AsRef<Path>>(file_path: P) -> Result<Self, PemError> {
+        Self::from_pem(File::open(file_path)?)
     }
 
     /// Creates an identity from a PEM certificate.
     #[cfg(feature = "pem")]
-    pub fn from_pem<R: std::io::Read>(pem_reader: R) -> Result<Self, PemError> {
-        let contents = pem_reader
-            .bytes()
-            .collect::<Result<Vec<u8>, std::io::Error>>()?;
-        let private_key = EcKey::private_key_from_pem(&contents)?;
-        Ok(Self::from_private_key(private_key))
+    pub fn from_pem<R: io::Read>(pem_reader: R) -> Result<Self, PemError> {
+        use sec1::{pem::PemLabel, EcPrivateKeyDocument};
+
+        const EC_PARAMETERS: &str = "EC PARAMETERS";
+        const SECP256K1: &[u8] = b"\x06\x05\x2b\x81\x04\x00\x0a";
+
+        let contents = pem_reader.bytes().collect::<Result<Vec<u8>, io::Error>>()?;
+
+        for pem in pem::parse_many(contents)? {
+            if pem.tag == EC_PARAMETERS && pem.contents != SECP256K1 {
+                return Err(PemError::ErrorStack(
```

Comment:
How about a dedicated PemError for this? It's not going to be quite correct to say "A key was rejected by k256: {0}", because k256 didn't reject it.  Something like PemError::UnsupportedKeyCurve, with an error message that we only support secp256k1 EC keys.

Reply:
Done. Worth noting that keys not generated by openssl (or generated without the params header) will still just say the key is malformed.